Support for Stash DIY backup

legal90 opened the pull request:
Implemented Stash DIY backup (#63)
More information about it: https://confluence.atlassian.com/display/STASH/Using+Stash+DIY+Backup

1) The following attributes have been moved from the namespace `['stash']['backup_client']` to the shared one `['stash']['backup']`:

* backup_path
* baseurl
* password
* user

I've tried to save backward compatibility by converting these attributes automatically. I've also added a warning message about it. This function (`check_for_old_attributes!`) can be removed in the one of the next releases.

2) Cron job configuration has been moved to the recipes: 'backup_client' and 'backup_diy'. 

So, the recipe 'backup_cron' is marked as deprecated, but it is still working and just the warning is shown. It could be removed in the one of the next releases

**Known restrictions for the restoring DIY backup:**
If you decide to restore your backup using `stash.diy-restore.sh` script, you should know that it might require more privileges from your database user.
For example, to upload PostgreSQL dump it could fail while attempting to create database, or add some relations, constraints, etc.
So that, it is recommended to use superuser credentials to restore your database backup.

Just let me know if you have any suggestions about this pull-request.
Thank you!

legal90 commented:
About Travic CI failure: 
Rubocop's warning does not relate to this changeset, so let's fix it separately.

linc01n commented:
Hi @legal90 ,

I have fixed the rubocop error and re-run travis test. Your PR now passed.

linc01n commented on recipes/backup_client.rb:
```diff
@@ -17,10 +18,27 @@
   to "#{node['stash']['home_path']}/backup-config.properties"
 end
+
+directory node.set['stash']['backup']['backup_path'] do
```
I think this should be `directory node['stash']['backup']['backup_path'] do`

legal90 replied:
Oh, sure, it should. I'll fix it

linc01n commented on recipes/backup_client.rb:
```diff
@@ -17,10 +18,27 @@
+
+directory node.set['stash']['backup']['backup_path'] do
+  owner node['stash']['user']
+  mode '0750'
```
If this one is the default path `/tmp`, then other users can't access tmpfs.

legal90 replied:
Do you mean that there should be `0755` ?

linc01n replied:
or change to default not to use `/tmp`? create a folder elsewhere? I understand the reason behind you set it to `0750`. Other user on the servers should not be able to pull all your backup.

legal90 replied:
Now I got it. Of course, we should not override permissions to "/tmp". 
So, we could just add a guard to this 'directory' resource:
```ruby
...
not_if { ::Dir.exists?(node['stash']['backup']['backup_path']) }
```

linc01n replied:
The backup archive is using `stash` as the owner. I think even in `/tmp` where others can see the archive file but they cannot open it. We are still safe enough. 

I think guarding `/tmp` from changing permission is :+1:

linc01n commented on recipes/backup_diy.rb:
```diff
@@ -0,0 +1,49 @@
+
+settings = Stash.settings(node)
+
+package 'jq'
```
`jq` is not in standard centos repo. It appears in debian/ubuntu/fedora only

legal90 replied:
Yes, you're right. `jq` is in EPEL repo only. What do you think about adding this line here?
```ruby
include_recipe 'yum-epel' if node['platform_family'] == 'rhel'
```
P.s. And add `yum-epel` dependensy to `metadata.rb`

linc01n replied:
:+1:

legal90 commented:
@linc01n I've added some changes by your recommendation. Please, review

legal90 commented:
I've rebased my branch to the current HEAD (+ squashed some commits).

linc01n commented:
Thanks !  I am able to do a fresh install in centos 6. Trying to test an upgrade scenario from backup client to diy.

linc01n commented on templates/default/backup-diy-vars.sh.erb:
```diff
@@ -0,0 +1,71 @@
+
+# The path to working folder for the backup
+STASH_BACKUP_ROOT=<%= @backup_diy['temp_path'] %>
+STASH_BACKUP_DB=${STASH_BACKUP_ROOT}/stash-db/
```
For mysql, the backup will fail since `mysqldump` will dump to a file but not a directory.

Should we add a condition check?

```erb
<% if node['stash']['database']['type'] == 'mysql' %>
STASH_BACKUP_DB=${STASH_BACKUP_ROOT}/stash-db
<% else %>
STASH_BACKUP_DB=${STASH_BACKUP_ROOT}/stash-db/
<% end %>
```

legal90 replied:
Seems like you're right. I don't know is it better to workaround this here, or we should fix it in "stash-diy-backup" scripts.
I'll try to investigate it soon.

linc01n replied:
Thanks.
Main problem here is if we use default db which is mysql. Turn on cron backup diy. Then this backup script will crash and won't recover itself. The whole instance will be stuck in backing up screen and user can't cancel the backup since they won't have the token.

legal90 replied:
Sorry for a delay. I've fixed the behavior by the similar way as you suggested above: 5d741cfda2ec7c35403e1fcee63a445a71785c3b

A've also rebased my branch to fix merge conflicts.

legal90 commented:
@linc01n Sorry for a delay. I've fixed the behavior by the similar way as you suggested before: 5d741cf

A've also rebased my branch to fix merge conflicts.

linc01n commented:
According to this 
https://bitbucket.org/atlassianlabs/atlassian-stash-diy-backup/pull-request/7/stashdev-7993-added-diy-restore/diff

it is safe to exclude mysql at all. Even you backup it, you can't restore it using diy method.

Let me do more testing later today. Then I will merge this in.

legal90 commented:
We can add a notice to README, describing that diy backup is provided "as is" and currently restore is available only for PostgreSQL (and put this link).

linc01n commented:
> We can add a notice to README, describing that diy backup is provided "as is" and currently restore is available only for PostgreSQL (and put this link).

Do you think we should put this in v3 or wait v4 when we support postgresql as the default database?

/cc @bflad

legal90 commented:
Hm... Then I think it will be better to put this and release in the scope of v4.

linc01n commented:
Merged! Thanks!
